larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -68,123 +68,88 @@ DenseElementsAttr Bitpack(PatternRewriter& builder, Attribute x) {
-
-    if (intermediate_maxpool_op &&
-        intermediate_maxpool_op.fused_activation_function() != "NONE") {
-      return failure();
-    }
-
-    auto inner_bconv_op =
-        intermediate_maxpool_op
-            ? dyn_cast_or_null<TF::LceBconv2dOp>(
-                  intermediate_maxpool_op.input().getDefiningOp())
-            : dyn_cast_or_null<TF::LceBconv2dOp>(
-                  outer_bconv_op.input().getDefiningOp());
-
-    if (inner_bconv_op && inner_bconv_op.padding() == "SAME" &&
-        inner_bconv_op.pad_values() != 1) {
-      // We don't return `failure()` here because we still want to match the
-      // pattern `binary maxpool -> binary conv`.
-      inner_bconv_op = nullptr;
-    }
-
-    if (!inner_bconv_op && !intermediate_maxpool_op) return failure();
-
-    if (!outer_bconv_op.input().hasOneUse() ||
-        (intermediate_maxpool_op && inner_bconv_op &&
-         !intermediate_maxpool_op.input().hasOneUse())) {
-      return failure();
-    }
+/**
+ * Returns `false` if some assumption does not hold (the consumer should then
+ * consider the rewrite pattern match to have failed).
+ */
+bool getBitpackedType(PatternRewriter& rewriter, ShapedType existing_type,
+                      RankedTensorType& bitpacked_type) {
+  if (existing_type.getElementType().isInteger(32)) return false;
+
+  const auto existing_shape = existing_type.getShape();
+
+  if (existing_shape.size() != 4) return false;
+
+  const auto channels = existing_shape[3];
+  const auto packed_channels = (channels + 32 - 1) / 32;
+  bitpacked_type = RankedTensorType::get({existing_shape[0], existing_shape[1],
+                                          existing_shape[2], packed_channels},
+                                         rewriter.getIntegerType(32));
+  return true;
+}
 
-    constexpr int bitwidth = 32;  // We use 32-bit bitpacking.
+template <typename BinaryOp>
+struct SetBitpackedActivations : public OpRewritePattern<BinaryOp> {
+  using OpRewritePattern<BinaryOp>::OpRewritePattern;
 
-    /*********************
-     * Compute new types *
-     *********************/
+  LogicalResult matchAndRewrite(BinaryOp outer_binary_op,
+                                PatternRewriter& rewriter) const override {
+    Operation* input_op = outer_binary_op.input().getDefiningOp();
 
-    RankedTensorType new_inner_bconv_op_type;
-    RankedTensorType new_intermediate_maxpool_op_type;
+    // If the input op has more than one use, we can't apply an optimisation.
+    if (!input_op || !input_op->hasOneUse()) return failure();
 
-    // If applicable, compute the new type for the inner bconv op.
+    // Try and match `input_op` to a binary convolution.
+    auto inner_bconv_op = dyn_cast_or_null<TF::LceBconv2dOp>(input_op);
     if (inner_bconv_op) {
-      ShapedType inner_bconv_op_type =
-          inner_bconv_op.getType().cast<ShapedType>();
+      if (inner_bconv_op.padding() == "SAME" &&
+          inner_bconv_op.pad_values() != 1) {
```

Comment:
I guess in the future (maybe if we have our own dialect) we could generalize and simplify those checks by using [`OpInterfaces`](https://mlir.llvm.org/docs/Interfaces/) similar to how [TFLite handles setting a channel axis](https://github.com/tensorflow/tensorflow/blob/3d9ec6298a5d167c93e260605d3d2957d294fcb2/tensorflow/compiler/mlir/lite/ir/tfl_ops.td#L780-L783) for different ops.